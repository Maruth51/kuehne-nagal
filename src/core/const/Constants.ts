import { TProductCategories } from "core/types/Product.types";

export const PRODUCT_CATEGORIES: TProductCategories[] = [
  'Software Development',
  'Daily Business',
  'Graphic Editors',
  'Text Editors',
  'Management Tools'
]