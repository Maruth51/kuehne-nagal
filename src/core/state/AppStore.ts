import { fetchProducts } from "core/services/productDataService";
import { IProduct, TProductCategories } from "core/types/Product.types";
import { makeAutoObservable } from "mobx";


class AppStore {
  activeTab = 'product'
  selectedCategories: TProductCategories[] = []
  searchString = '';
  selectedProduct: IProduct | null = null;
  products: IProduct[] = []
  constructor() {
    makeAutoObservable(this)
  }

  search(s: string) {
    this.searchString = s
  }

  selectProduct(product: IProduct) {
    this.selectedProduct = product
  }

  get filteredProducts(): IProduct[] {
    if(this.searchString === '') {
      return []
    }
    return this.products.filter((product : IProduct) => {
      // if not category selected, by default all category included
      let isProductCategorySelected = true;
      if(!!this.selectedCategories.length) {
        isProductCategorySelected = this.selectedCategories.includes(product.category)
      }
      return isProductCategorySelected && product.productName.toLowerCase().includes(this.searchString.trim().toLowerCase())
    })
  }

  toggleSelectedCategory(category: TProductCategories) {
    const categoryIdx = this.selectedCategories.indexOf(category)
    if (categoryIdx === -1) {
      this.selectedCategories.push(category)
    } else {
      this.selectedCategories.splice(categoryIdx, 1)
    }
  }

  async fetchProducts() {
    this.products = await fetchProducts()
  }
}

export default AppStore;