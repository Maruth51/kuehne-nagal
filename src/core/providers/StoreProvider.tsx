import React from "react";
import AppStore from "core/state/AppStore";

const StoreContext = React.createContext<AppStore>(new AppStore())

export const useStore = () => React.useContext(StoreContext);

export const StoreProvider: React.FC<{ children: React.ReactNode, store: AppStore}> = ({ children, store }) => {
  return (
    <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
  );
};