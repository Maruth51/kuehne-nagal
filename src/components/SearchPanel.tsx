import { PRODUCT_CATEGORIES } from "core/const/Constants";
import { useStore } from "core/providers/StoreProvider";
import { IProduct, TProductCategories } from "core/types/Product.types";
import { observer } from "mobx-react-lite";
import React from "react";
import styles from 'styles/SearchPanel.module.scss';
import Product from "./Product";


const SearchPanel = () => {
  const store = useStore()

  const handleCategoryClick = (category: TProductCategories) => () => {
    store.toggleSelectedCategory(category)
  }

  const handleSearch = (e : React.ChangeEvent<HTMLInputElement>) => {
    store.search(e.target.value)
  }

  const handleProductSelect = (product: IProduct) => () => {
    store.selectProduct(product)
  }

  return <>
  <div  data-testid="search-panel" className={styles.panel}>
    <p className={styles.title}> Iâ€™m looking for... </p>
    <div className={styles.filter}>

      <ul className={styles.categoryContainer}>
        {PRODUCT_CATEGORIES.map((category, idx) => {
          return <li
            key={idx}
            className={styles.category}
          >
            <input
              name={category}
              type='checkbox'
              onChange={handleCategoryClick(category)}
              checked={store.selectedCategories.includes(category)}
              id={'category-' + idx}
            />
            <label htmlFor={'category-' + idx} className="label">
              {category}
            </label>
          </li>
        })}
      </ul>
      <div className={styles.inputContainer}>

        <div className={styles.searchBox}>
          <img src="/search_icon.svg" alt="search icon" />
          <input type="text" name="search-input" onChange={handleSearch} placeholder='Type here...'/>
        </div>
      </div>
    </div>
  </div>
  {
    store.filteredProducts.map((product) => {
      return <Product product={product} onSelect={handleProductSelect(product)}/>
    })
  }
  </>
}

export default observer(SearchPanel);