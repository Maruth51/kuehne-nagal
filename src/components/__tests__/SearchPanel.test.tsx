import { fireEvent, render, screen } from "@testing-library/react"
import SearchPanel from "components/SearchPanel"
import { PRODUCT_CATEGORIES } from "core/const/Constants"
import { StoreProvider } from "core/providers/StoreProvider"
import AppStore from "core/state/AppStore"

describe('Search panel component', () => {
  let store: AppStore
  
  beforeEach(() => {
    store = new AppStore()
    store.fetchProducts()
  })

  const setup = () => {
   return render(<StoreProvider store={store}>
      <SearchPanel />
    </StoreProvider>)
  }
  

  test('renders on initial', () => {
    setup()
    const searchPanel = screen.getByTestId('search-panel')
    expect(searchPanel).toBeInTheDocument()
  })

  test('renders all child elements', () => {
    setup()
    const searchPanel = screen.getByTestId('search-panel')
    expect(searchPanel).toMatchSnapshot('SearchPanelComponent')
  })

  test('search for products', () => {
    setup()
    const searchFor  = 'test search'
    const searchInputBox = screen.getByPlaceholderText<HTMLInputElement>('Type here...')
    fireEvent.change(searchInputBox, {target: {value: searchFor}})
    expect(searchInputBox.value).toEqual(searchFor)
  })

  test.each(PRODUCT_CATEGORIES)('select product category - %s', (category) => {
    setup()
    const categoryCheckBox = screen.getByLabelText<HTMLInputElement>(category)
    fireEvent.click(categoryCheckBox)
    expect(store.selectedCategories).toContain(category)
  })

  test.each(PRODUCT_CATEGORIES)('remove product category selection - %s', (category) => {
    setup()
    const categoryCheckBox = screen.getByLabelText<HTMLInputElement>(category)
    // Checked on first click
    fireEvent.click(categoryCheckBox)

     // UnChecked on next click
     fireEvent.click(categoryCheckBox)
    expect(store.selectedCategories).not.toContain(category)
  })
})