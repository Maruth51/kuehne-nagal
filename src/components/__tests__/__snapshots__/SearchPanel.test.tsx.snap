// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search panel component renders all child elements: SearchPanelComponent 1`] = `
<div
  class="panel"
  data-testid="search-panel"
>
  <p
    class="title"
  >
     Iâ€™m looking for... 
  </p>
  <div
    class="filter"
  >
    <ul
      class="categoryContainer"
    >
      <li
        class="category"
      >
        <input
          id="category-0"
          name="Software Development"
          type="checkbox"
        />
        <label
          class="label"
          for="category-0"
        >
          Software Development
        </label>
      </li>
      <li
        class="category"
      >
        <input
          id="category-1"
          name="Daily Business"
          type="checkbox"
        />
        <label
          class="label"
          for="category-1"
        >
          Daily Business
        </label>
      </li>
      <li
        class="category"
      >
        <input
          id="category-2"
          name="Graphic Editors"
          type="checkbox"
        />
        <label
          class="label"
          for="category-2"
        >
          Graphic Editors
        </label>
      </li>
      <li
        class="category"
      >
        <input
          id="category-3"
          name="Text Editors"
          type="checkbox"
        />
        <label
          class="label"
          for="category-3"
        >
          Text Editors
        </label>
      </li>
      <li
        class="category"
      >
        <input
          id="category-4"
          name="Management Tools"
          type="checkbox"
        />
        <label
          class="label"
          for="category-4"
        >
          Management Tools
        </label>
      </li>
    </ul>
    <div
      class="inputContainer"
    >
      <div
        class="searchBox"
      >
        <img
          alt="search icon"
          src="/search_icon.svg"
        />
        <input
          name="search-input"
          placeholder="Type here..."
          type="text"
        />
      </div>
    </div>
  </div>
</div>
`;
